src/standard.ts

```diff
@@ -23,58 +23,71 @@ export class StandardMerkleTree<T extends any[]> {
     private readonly tree: Bytes[],
     private readonly values: { value: T, treeIndex: number }[],
     private readonly leafEncoding: string[],
+    private readonly hashFn: HashFn,
   ) {
     this.hashLookup =
       Object.fromEntries(values.map(({ value }, valueIndex) => [
-        hex(standardLeafHash(value, leafEncoding)),
+        hex(leafHash(value, leafEncoding, this.hashFn)),
         valueIndex,
       ]));
   }
 
   static of<T extends any[]>(values: T[], leafEncoding: string[], options: MerkleTreeOptions = {}) {
+    const hashFn = options.hashFn ?? defaultOptions.hashFn;
     const sortLeaves = options.sortLeaves ?? defaultOptions.sortLeaves;
 
-    const hashedValues = values.map((value, valueIndex) => ({ value, valueIndex, hash: standardLeafHash(value, leafEncoding) }));
+    const hashedValues = values.map((value, valueIndex) => ({
+      value,
+      valueIndex,
+      hash: leafHash(value, leafEncoding, hashFn),
+    }));
 
     if (sortLeaves) {
       hashedValues.sort((a, b) => compareBytes(a.hash, b.hash));
     }
 
-    const tree = makeMerkleTree(hashedValues.map(v => v.hash));
+    const tree = makeMerkleTree(
+      hashedValues.map((v) => v.hash),
+      (a, b) => hashPair(a, b, hashFn)
+    );
 
     const indexedValues = values.map(value => ({ value, treeIndex: 0 }));
     for (const [leafIndex, { valueIndex }] of hashedValues.entries()) {
       indexedValues[valueIndex]!.treeIndex = tree.length - leafIndex - 1;
     }
 
-    return new StandardMerkleTree(tree, indexedValues, leafEncoding);
+    return new StandardMerkleTree(tree, indexedValues, leafEncoding, hashFn);
   }
-
-  static load<T extends any[]>(data: StandardMerkleTreeData<T>): StandardMerkleTree<T> {
+ 
+  static load<T extends any[]>(data: StandardMerkleTreeData<T>, hashFn?: HashFn): StandardMerkleTree<T> {
     if (data.format !== 'standard-v1') {
       throw new Error(`Unknown format '${data.format}'`);
     }
     return new StandardMerkleTree(
       data.tree.map(hexToBytes),
       data.values,
       data.leafEncoding,
+      hashFn ?? standardHash,
```

Comment:
This is an issue. If the tree was built using a non standard hash, an dumped() to a file, how do you know how to load it? What happens if its loaded using no hashFn, or the wrong hashFn?

When we built this library, we felt like dump/load was essential. Using custom hashes breaks some assumption about the ability to dump/load seemlesly.

Reply:
The dump should contain a line that says `"hash": "custom"`. `load` should have a `hashFn` argument that must be present if the dump contains this line. The user is responsible for passing in the correct hash, although we could do a minimal sanity check at load time to error early (i.e., try to hash something and see that it matches an expected value).

Reply:
Something like this ?
https://github.com/OpenZeppelin/merkle-tree/blob/5e509046ce21d6c34614c5c5d3fee8f0f5492fcb/src/simple.ts#L77-L94